codegen/kuberesource/crd.go

```diff
@@ -1,26 +1,108 @@
 package kuberesource
 
 import (
+	"encoding/json"
 	"fmt"
 	"strings"
 
+	"cuelang.org/go/encoding/openapi"
 	"github.com/gertd/go-pluralize"
-
+	"github.com/rotisserie/eris"
 	"github.com/solo-io/skv2/codegen/model"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
 	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Mapping from protobuf message name to OpenApi schema
+type OpenApiSchemas map[string]*openapi.OrderedMap
+
 // Create CRDs for a group
-func CustomResourceDefinitions(group model.Group) []metav1.Object {
-	var objects []metav1.Object
+func CustomResourceDefinitions(group model.Group, oapiSchemas OpenApiSchemas) (objects []metav1.Object, err error) {
 	for _, resource := range group.Resources {
-		objects = append(objects, CustomResourceDefinition(resource))
+
+		var validationSchema *apiextv1beta1.CustomResourceValidation
+		if group.RenderValidationSchemas {
+			validationSchema, err = constructValidationSchema(resource, oapiSchemas)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		objects = append(objects, CustomResourceDefinition(resource, validationSchema))
+	}
+	return objects, nil
+}
+
+func constructValidationSchema(resource model.Resource, oapiSchemas OpenApiSchemas) (*apiextv1beta1.CustomResourceValidation, error) {
+	specSchema, err := getJsonSchema(resource, resource.Spec.Type.Name, oapiSchemas)
+	if err != nil {
+		return nil, eris.Wrapf(err, "constructing spec validation schema for Kind %s", resource.Kind)
+	}
+	statusSchema, err := getJsonSchema(resource, resource.Status.Type.Name, oapiSchemas)
+	if err != nil {
+		return nil, eris.Wrapf(err, "constructing status validation schema for Kind %s", resource.Kind)
+	}
+	return &apiextv1beta1.CustomResourceValidation{
+		OpenAPIV3Schema: &apiextv1beta1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]apiextv1beta1.JSONSchemaProps{
+				"spec":   *specSchema,
```

Comment:
This works because of the expanded references?

Reply:
Yup, exactly, the generated openapi schemas are fully self contained because `ExpandReferences: true` on the cuelang cfg